### ISSUE TITLE
OC Circuit Programmer

### ISSUE THREAD

**Issue by Victorious3** [Victorious3/Integrated-Circuits#11]
Just like the ComputerCraft program, but for OpenComputers.

**Comment by skyem123** [comment]
I am working on it.

**Comment by skyem123** [comment]
Hrm, one big problem is the fact that ComputerCraft has APIs that OpenComputers doesn't.

**Comment by Thog** [comment]
You can write it them :p

**Comment by skyem123** [comment]
Yeah, that's what I'm planning to do.
- [ ] Make GUI lib for graphical stuff
- [ ] Make final program

**Comment by Victorious3** [comment]
At least you can get rid of my dirty hack overriding window's methods to implement a scissor test like that.

**Comment by skyem123** [comment]
Todo list for the GUI API
- [x] Colour API
- [x] Character / String Ouput
- [x] Simple Line Drawing 
- [ ] Diagonal Line Drawing
- [ ] Simple Rectangle Drawing
- [ ] Basic Window Functions
- [ ] Basic Page System
- [ ] Scrollable Page
- [ ] Menus
- [ ] Basic Dialogue Box
- [ ] Controls
- [ ] Dialogue Box With Controls

**Comment by skyem123** [comment]
Todo list for the program
- [ ] Need GUI Lib
- [ ] Finish Todo list

**Comment by skyem123** [comment]
Should I make a gist to put my WIP of my programing?

**Comment by Victorious3** [comment]
You should and, btw, I hardcoded all the controls using a few simple mouse intersection tests. It is nothing you'd need a special library for. It isn't meant to be complete, as long as it works it is fine.

**Comment by Techokami** [comment]
Imho it would be nice to have a full API for other users to make GUI programs in OC, too, but that can be for after getting the IC program finished.

Sent from my Windows Phone
________________________________

**Comment by skyem123** [comment]
Yeah, I'm trying to make the GUI API as flexible as possible for that reason.

**Comment by skyem123** [comment]
Gah. I'm too dumb to understand diagonal lines, so I'm giving up.

**Comment by GopherAtl** [comment]
at the risk of self-promotion, there is my gui library, GML, over in [OpenPrograms](https://github.com/OpenPrograms/Gopher-Programs)  You could use that rather than start from scratch on the gui stuff. The repo wiki is pretty thorough, and there's sample programs in the gml/bin directory, as well as the [CompViewer](https://github.com/OpenPrograms/Kenny-Programs/tree/master/CompViewer) program Kenny made, which uses it

**Comment by skyem123** [comment]
Well, here is my progress so far: https://gist.github.com/skyem123/529593c86a36f1d38a83

**Comment by skyem123** [comment]
@GopherAtl, your GUI lib seems very large and complex, though I might be able to take some code from it. The one problem is licensing (sadly, I am kinda scared of this stuff), and so will you allow me to use your code in CC0-licensed code?

**Comment by GopherAtl** [comment]
there's not a licence file, and I suppose there should be, but it's wtfpl in the truest sense, so do wtf you want XD

It definitely is rather large and complex, though, and may not be well-suited to this project. If you have hopes of being able to pluck out and use a few bits without using the whole thing, you might find that difficult.

**Comment by skyem123** [comment]
Heh, thank you! :smile:

**Comment by skyem123** [comment]
Okay, I'm probably going to have to continue working on this...

**Comment by skyem123** [comment]
Gah. I'm honestly not sure what to do with this...

**Comment by skyem123** [comment]
@Victorious3, do you still want me to do this?

**Comment by Victorious3** [comment]
Sure.

**Comment by skyem123** [comment]
Okay... @Victorious3, I have run out of motivation to do this, and you seem to know a lot more about programming (especially graphics) than me... can you please do this?

**Comment by Victorious3** [comment]
Nope. https://twitter.com/VicNightfall/status/657558160740327424

**Comment by skyem123** [comment]
Please... You can do this better and quicker than I can...

**Comment by Pyeroh** [comment]
Eyh @skyem123, don't give up. What do you need to do ? Programming is my job, and I'd be glad to give a hand. Just make a list of what you really need, and if it's not me, anyone would be able to give you bits of code to close this issue :D

**Comment by skyem123** [comment]
@Pyeroh, to be honest, @Victorious3 is considering dropping the feature that this relies on, so I don't want to work on something that would be useless...
and I also haven't really designed what I really wanted...

**Comment by Victorious3** [comment]
No, I'm not.

**Comment by skyem123** [comment]
@Victorious3, also you need to fix the CC program, it's rendering is slightly broken on the latest ComputerCraft...

**Comment by skyem123** [comment]
@Victorious3, if I write an minimal CC emulator, will you fix bugs in your CC program and make it less horrible?

**Comment by Victorious3** [comment]
@skyem123 Mkay
